src/main/java/com/gregtechceu/gtceu/common/machine/multiblock/electric/research/ComputationProviderMachine.java

```diff
@@ -0,0 +1,169 @@
+        seen.add(this);
+        if (maxCWUt == 0)
+            return maxCWUt = machineCallback("getMaxCWUt", null, 0);
+        else return maxCWUt;
+    }
+
+    @Override
+    public boolean canBridge(@NotNull Collection<IOpticalComputationProvider> seen) {
+        seen.add(this);
+        if (!isFormed()) return true;
+        if (canBridge == null) return canBridge = machineCallback("canBridge", null, true);
+        return canBridge;
+    }
+
+    public void tick() {
+        totalCWU -= allocatedCWUt;
+        lastAllocatedCWUt = String.valueOf(allocatedCWUt);
+        if(getRecipeLogic().isSuspend()){
+            allocatedCWUt = 0;
+            canProvideCWUt=false;
+            return;
+        }else{
+            canProvideCWUt=true;
+        }
+        if(allocatedCWUt!=0){
+            getRecipeLogic().setStatus(RecipeLogic.Status.WORKING);
+            allocatedCWUt = 0;
+        }else{
+            getRecipeLogic().setStatus(RecipeLogic.Status.IDLE);
+        }
+        if (totalCWU < getMaxCWUt()) {
+            totalCWU += machineCallback("requestCWUt", null, 0);
+            maxCWUt = 0;
+        }
+    }
+
+    @Override
+    public void onLoad() {
+        super.onLoad();
+        if (getLevel() instanceof ServerLevel serverLevel) {
+            serverLevel.getServer().tell(new TickTask(0, this::updateTickSubscription));
+        }
+    }
+
+    @Override
+    public void onUnload() {
+        super.onUnload();
+        if (tickSubs != null) {
+            tickSubs.unsubscribe();
+            tickSubs = null;
+        }
+    }
+
+    protected void updateTickSubscription() {
+        if (isFormed) {
+            tickSubs = subscribeServerTick(tickSubs, this::tick);
+        } else if (tickSubs != null) {
+            tickSubs.unsubscribe();
+            tickSubs = null;
+        }
+    }
+
+    @Override
+    public void onStructureFormed() {
+        super.onStructureFormed();
+        if (getLevel() instanceof ServerLevel serverLevel) {
+            serverLevel.getServer().tell(new TickTask(0, this::updateTickSubscription));
+        }
+    }
+
+    @Override
+    public void onChanged() {
+//        super.onChanged();
+        GTCEu.LOGGER.debug("change");
+        maxCWUt = 0;
+    }
+
+    @Override
+    public void addDisplayText(List<Component> textList) {
+        MultiblockDisplayText.builder(textList, isFormed())
+            .setWorkingStatus(true, allocatedCWUt > 0) // transform into two-state system for
+            // display
+            .setWorkingStatusKeys(
+                "gtceu.multiblock.idling",
+                "gtceu.multiblock.idling",
+                "gtceu.multiblock.data_bank.providing")
+            .addCustom(tl -> {
+                if (isFormed()) {
+                    // Provided Computation
+
+                    Component cwutInfo = Component.literal(
+                            lastAllocatedCWUt + " / " + getMaxCWUt() + " CWU/t")
```

Comment:
we have lang keys for these IIRC.

Reply:
this is copy from hpca

Reply:
oh then I forgot to add that there. oops.